zkdb/src/storage/trees/merkle/MerkleTreeJSON.ts

```diff
@@ -0,0 +1,120 @@
+import { CID } from 'multiformats/cid';
+import { Field } from 'snarkyjs';
+import * as IPFS from 'ipfs-core';
+import { BaseMerkleTree, NodesMap } from './BaseMerkleTree.js';
+
+/**
+ * MerkleTreeJSON class is an implementation of the BaseMerkleTree that
+ * stores the Merkle tree nodes in IPFS as a JSON file.
+ */
+export default class MerkleTreeJSON extends BaseMerkleTree {
+  private storage: IPFS.IPFS;
+  private nodesFileCID: CID;
+
+  /**
+   * Creates a new MerkleTreeJSON instance.
+   *
+   * @param height The height of the Merkle tree.
+   * @param storage The IPFS storage instance to store the nodes.
+   * @param root The root CID of the Merkle tree (optional, defaults to null).
+   */
+  constructor(
+    height: number,
+    storage: IPFS.IPFS,
+    nodesFileCID: CID | null = null
+  ) {
+    super(height);
+    this.storage = storage;
+    if (nodesFileCID) this.nodesFileCID = nodesFileCID;
+  }
+
+  public async getNode(level: number, index: bigint): Promise<Field> {
+    if (this.isEmpty()) {
+      return this.zeroes[level];
+    }
+
+    const nodes = await this.getNodes();
+    return this.getNodeOrZero(level, index, nodes);
+  }
+
+  public async getRoot(): Promise<Field> {
+    if (this.isEmpty()) {
+      return this.zeroes[this.height - 1];
+    }
+    const nodes = await this.getNodes();
+
+    const value = nodes[this.height - 1]['0'];
+
+    return Field(value);
+  }
+
+  protected async saveLeaf(nodesMap: NodesMap): Promise<void> {
+    const currentNodesMap = await this.getNodes();
+
+    if (
+      Object.keys(currentNodesMap).length !== Object.keys(nodesMap).length &&
+      Object.keys(currentNodesMap).length !== 0
+    ) {
+      throw new Error(`the length of trees are not equal`);
+    }
+
+    for (const level in nodesMap) {
+      const nodesAtLevel = nodesMap[level];
+
+      for (const [nodeIndex, nodeValue] of Object.entries(nodesAtLevel)) {
+        if (!currentNodesMap[level]) {
+          currentNodesMap[level] = {};
+        }
+        currentNodesMap[level][nodeIndex] = nodeValue;
+      }
+    }
+
+    await this.saveNodes(currentNodesMap);
+  }
+
+  // eslint-disable-next-line no-unused-vars
+  protected async getMerklePath(leafIndex: bigint): Promise<NodesMap> {
+    return await this.getNodes();
+  }
+
+  protected async retrieveNodes(): Promise<NodesMap> {
+    if (!this.nodesFileCID) {
+      return {};
+    }
+    const chunks = [];
+
+    for await (const chunk of this.storage.cat(this.nodesFileCID)) {
+      chunks.push(chunk);
+    }
+
+    const jsonString = Buffer.concat(chunks).toString();
+    const jsonObject: NodesMap = JSON.parse(jsonString);
```

Comment:
I would prefer, to use BSON instead of JSON since it's more friendly with binary and also the serialization/deserialization much more better than JSON.